src/Init/Core.lean

```diff
@@ -1680,40 +1680,92 @@ So, you are mainly losing the capability of type checking your development using
-(which also simplifies up to idempotence when available).
+`IdempotentOp op` indicates `op` is an idempotent binary operation.
+i.e. `a ∘ a = a`.
 -/
-class IsIdempotent {α : Sort u} (op : α → α → α) where
+class IdempotentOp (op : α → α → α) : Prop where
   /-- An idempotent operation satisfies `a ∘ a = a`. -/
   idempotent : (x : α) → op x x = x
 
 /--
-`IsNeutral op e` says that `e` is a neutral operation for `op`,
-i.e. `a ∘ e = a = e ∘ a`. It is used by the `ac_rfl` tactic
-(which also simplifies neutral elements when available).
+`LeftIdentify op o` indicates `o` is a left identity of `op`.
+
+This class does not require a proof that `o` is an identity, and
+is used primarily for infering the identity using class resoluton.
 -/
-class IsNeutral {α : Sort u} (op : α → α → α) (neutral : α) where
-  /-- A neutral element can be cancelled on the left: `e ∘ a = a`. -/
-  left_neutral : (a : α) → op neutral a = a
-  /-- A neutral element can be cancelled on the right: `a ∘ e = a`. -/
-  right_neutral : (a : α) → op a neutral = a
+class LeftIdentity (op : α → β → β) (o : outParam α) : Prop where
 
-end Lean
+/--
+`LawfulLeftIdentify op o` indicates `o` is a verified left identity of
+`op`.
+-/
+class LawfulLeftIdentity (op : α → β → β) (o : outParam α) extends LeftIdentity op o : Prop where
+  /-- Left identity `o` is an identity. -/
+  left_id : ∀ a, op o a = a
+
+/--
+`RightIdentify op o` indicates `o` is a right identity `o` of `op`.
+
+This class does not require a proof that `o` is an identity, and is used
+primarily for infering the identity using class resoluton.
+-/
+class RightIdentity (op : α → β → α) (o : outParam β) : Prop where
+
+/--
+`LawfulRightIdentify op o` indicates `o` is a verified right identity of
+`op`.
+-/
+class LawfulRightIdentity (op : α → β → α) (o : outParam β) extends RightIdentity op o : Prop where
+  /-- Right identity `o` is an identity. -/
+  right_id : ∀ a, op a o = a
+
+/--
+`Identity op o` indicates `o` is a left and right identity of `op`.
+
+This class does not require a proof that `o` is an identity, and is used
+primarily for infering the identity using class resoluton.
+-/
+class Identity (op : α → α → α) (o : outParam α) extends LeftIdentity op o, RightIdentity op o : Prop where
+
+/--
+`LawfulIdentity op o` indicates `o` is a verified left and right
+identity of `op`.
+-/
+class LawfulIdentity (op : α → α → α) (o : outParam α) extends Identity op o, LawfulLeftIdentity op o, LawfulRightIdentity op o : Prop
+
+/--
+`LawfulCommIdentity` can simplify defining instances of LawfulIdentity
+on commutative functions by requiring only a left or right identity
+proof.
+
+This class is intended for simplifying defining instances of
+`LawfulIdentity` and functions needed commutative operations with
+identity should just add a `LawfulIdentity constraint.
+-/
+class LawfulCommIdentity (op : α → α → α) (o : outParam α) [hc : Commutative op] extends LawfulIdentity op o : Prop where
```

Comment:
If this is for defining instances, perhaps making it also extend `Commutative op` would be useful. Is there a reason it's an argument instead?

Reply:
The one downside of doing that is that it also makes the class convenient for adding commutativity and identity constraints as well: `def foo [LawfulCommIdentity op o] -> ...` which is not the preferred form.

I'm inclined to just leave this as is for now.

Reply:
One more thought here: if you made this be a `structure`, you could do `attribute [instance] LawfulCommIdentity.toCommutative LawfulCommIdentity.LawfulIdentity`, and that would protect against it being used for an instance argument and it would give the benefits of it being a `class` (though it'd be worth checking that `LawfulCommIdentity.mk` has its `to*` fields be instance arguments once it's switched to `structure` -- I don't know that off the top of my head).

Reply:
If I try this:

```lean
structure LawfulCommIdentity (op : α → α → α) (o : α) extends Commutative op, LawfulIdentity op o : Prop where
  left_id a := Eq.trans (comm o a) (right_id a)
  right_id a := Eq.trans (comm a o) (left_id a)

attribute [instance] LawfulCommIdentity.toLawfulIdentity
```

I get the error:
```
instance does not provide concrete values for (semi-)out-params
  @LawfulIdentity α op ?o
```

I'm in favor of leaving this alone and merging as is, but alternatively I'm open to dropping `LawfulCommIdentity` entirely.  I think the benefits are pretty minor and I need to prioritize other topics.

Reply:
Thanks for checking (and in retrospect that was a silly suggestion, sorry).

While `LawfulCommIdentity` is an unfamiliar design (I don't think mathlib would have it), so it was worth thinking about some more (as it will probably be here for years if it makes it into core), I think I'm happy enough with this PR as it is. I'll look through once more.

Reply:
Thanks for the feedback though.  I appreciate it.